accord-core/src/main/java/accord/coordinate/Invalidate.java

```diff
@@ -151,12 +139,12 @@ private void invalidate()
                     {
                         RecoverWithRoute.recover(node, ballot, txnId, route, callback);
```

Comment:
From the description we should only ever invalidate on a random shard once, we should only attempt to invalidate on the home shard once, and then finally we might invalidate across all shards.

Maybe the invariant is a limit on not invalidating a txn using the same inputs twice? If we fail to invalidate it is because we switch to recovery or invalidating on a different set of shards.

Reply:
Yes, I think that would be a valid invariant. I'm not sure of a good way to enforce that, though? We could propagate the parameters most recently used for `Invalidate` (that are `@Nullable`) and ensure we are using strictly more information each time?

Reply:
Yeah. As long as we catch it eventually it's fine if we repeat a few useless attempts and converge on the maximum amount of what we can provide to invalidate. It's mostly just to increase visibility when the logic is wrong.

Reply:
Okay, in the process of addressing this I changed the semantics slightly. We now only reach `Success` in `InvalidationTracker` if we can reject the fast path OR have promises at every shard. This permits us to guarantee forward progress by only passing forwards the status we encountered in the first invalidation: if it was a recoverable status, then we fail if we cannot find enough information to recover. If it was not a recoverable status, we inform the recursive invalidation that it is safe to invalidate if it does not itself find a recoverable status, since its promises _combined with_ the failure to find enough information to recover _post_ these promises being made, guarantees that the original coordinator cannot reach a fast path decision (and also will abandon its coordination, as it has been preempted by us).

In principle we could weaken this, as _any_ shard promising _while also_ having insufficient information to recover would suffice. But it is simpler to just require that every shard we use for recovery is promised, so we know that if any shard we later contact has insufficient information then it's guaranteed to be safe to invalidate (unless we find a recoverable status, i.e. Accept or later, in which case we are guaranteed to find enough information to invoke recovery)

Reply:
Sounds good.